Fixing API errors

matheusmra opened the pull request:
I’ve just implemented the fix for the API error and tested it locally.
This update improves the endpoint by adding explicit handling for the most common failure scenarios. The following exceptions are now properly logged and mapped to meaningful HTTP responses:
```python
 except ValueError as e:
        logging.error(f"Validation error: {e}")
        raise HTTPException(status_code=400, detail=f"Validation Error: {e}")
    except json.JSONDecodeError as e:
        logging.error(f"JSON parsing error: {e}")
        raise HTTPException(status_code=400, detail=f"Invalid JSON format: {e}")
    except FileNotFoundError as e:
        logging.error(f"File not found: {e}")
        raise HTTPException(status_code=404, detail=f"File not found: {e}")
    except Exception as e:
        logging.exception(f"Internal Server Error: {e}")
        raise HTTPException(status_code=500, detail=f"Internal Server Error: {e}")
```
Let me know if you think there’s another scenario I should handle.

matheusmra committed:
Fixing API errors

ArthurCRodrigues commented on connectors/adapters/api/api_entrypoint.py:
```diff
@@ -93,7 +94,17 @@ async def grade_submission_endpoint(
+        raise HTTPException(status_code=404, detail=f"File not found: {e}")
     except Exception as e:
+        logging.exception(f"Internal Server Error: {e}")
         raise HTTPException(status_code=500, detail=f"Internal Server Error: {e}")
```
That's important.

ArthurCRodrigues reviewed (changes requested):
That's a nice way to start solving our problem but i think the optimal solution may encompass a more detailed investigation across our system in order to evaluate where this `'Nonetype' object  is not iterable` message is being created (the string itself is most likely not hardcoded in the project, it's probably generated by python itself). Also, if the exception happened inside the `/autograder` package the idea of to include the error as an `AutograderResponse` aswell, look at how the `autograder_facade` returns an error found:
```python
except Exception as e:
            # Catch any exception, log it, and return a failure response
            error_message = f"An unexpected error occurred during the grading process: {str(e)}"
            logger.error(error_message)
            logger.exception("Full exception traceback:")
            return AutograderResponse(status="fail", final_score=0.0, feedback=error_message)
```
And this seems not to be happening.


So i suggest a more detailed investigation over the scopes of our system that are related to throwing errors.

matheusmra committed:
Update api_entrypoint.py

matheusmra committed:
Update api_entrypoint.py

matheusmra commented:
Thanks for the heads-up, @ArthurCRodrigues !
After investigating the issue in the autograder flow, I identified two issues that were causing NoneType iteration errors in the API adapter.

### 1. AutograderResponse Default Value
The `test_report` field in `AutograderResponse` defaulted to `None` instead of an empty list:
```python
# BEFORE
test_report: Optional[List[TestResult]] = Field(default=None)
```

### 2. Exception Handler Missing Field
When exceptions occurred during grading, the exception handler in `autograder_facade.py` created an `AutograderResponse` without specifying `test_report`, causing it to use the `None` default, which later caused iteration errors in the API adapter.

## Solution

### Core Fixes

**1. Changed AutograderResponse model default** (`autograder/core/models/autograder_response.py`)
```python
# AFTER
test_report: List[TestResult] = Field(default_factory=list)
```

**2. Updated exception handler** (`autograder/autograder_facade.py`)
```python
return AutograderResponse(
    status="fail", 
    final_score=0.0, 
    feedback=error_message,
    test_report=[]  
)
```

I ran a lot of tests locally and didn't receive any NoneType errors after applying the fixes. 
If you run the updated version and encounter any error or unexpected behavior, please let me know

matheusmra committed:
Fixing API

ArthurCRodrigues reviewed (approved):
Awesome! Nailed it.
